File: z_lib/low-level-design/logging-service/logger.go
package logging

import "time"

const timeLayout = time.RFC3339Nano

type Logger struct {
	destinations []*Destination
}

func NewLogger(destinations []*Destination) *Logger {
	copied := make([]*Destination, len(destinations))
	copy(copied, destinations)
	return &Logger{destinations: copied}
}

func (logger *Logger) Log(level LogLevel, message string) {
	record := NewLogRecord(
		time.Now().UTC(),
		level,
		message,
		"goroutine",
	)

	for _, destination := range logger.destinations {
		destination.Write(record)
	}
}

func (logger *Logger) Debug(message string) {
	logger.Log(DEBUG, message)
}

func (logger *Logger) Info(message string) {
	logger.Log(INFO, message)
}

func (logger *Logger) Warn(message string) {
	logger.Log(WARN, message)
}

func (logger *Logger) Error(message string) {
	logger.Log(ERROR, message)
}

func (logger *Logger) Fatal(message string) {
	logger.Log(FATAL, message)
}

File: lld/logger.py
from logging_core import LogLevel
from appenders import LogAppender
from logging_core import LogMessage

class Logger:
    def __init__(self, name):
        self.name = name
        self.level = LogLevel.INFO
        self.appender:list[LogAppender] = []
        
    def add_appender(self, appendType: LogAppender):
        self.appender.append(appendType)
        return
    
    def log(self, level, msg: str):
        if self.level.value <= level.value:
            message = LogMessage(level, msg) 
            for app in self.appender:
                app.append(message)
            
    def debug(self, message: str):
        self.log(LogLevel.DEBUG, message)

    def info(self, message: str):
        self.log(LogLevel.INFO, message)
    
    def warning(self, message: str):
        self.log(LogLevel.WARNING, message)

    def error(self, message: str):
        self.log(LogLevel.ERROR, message)    
        
    def fatal(self, message: str):
        self.log(LogLevel.FATAL, message)